File: client/android/app/src/main/java/com/stonefive/chalkak/feature/display/DisplayViewModel.kt
package com.stonefive.chalkak.feature.display

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.stonefive.chalkak.ChalkakApplication
import com.stonefive.chalkak.core.ui.UiMessage
import com.stonefive.chalkak.domain.model.HomeFailure
import com.stonefive.chalkak.domain.model.HomeQuery
import com.stonefive.chalkak.domain.model.HomeResult
import com.stonefive.chalkak.domain.model.Post
import com.stonefive.chalkak.domain.model.PostContent
import com.stonefive.chalkak.domain.model.PostPage
import com.stonefive.chalkak.domain.model.PostSort
import com.stonefive.chalkak.domain.repository.PostRepository
import java.time.LocalDate
import java.time.ZoneId
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class DisplayViewModel(
    private val repository: PostRepository,
    private val initialDate: LocalDate? = null,
    private val dateProvider: () -> LocalDate = { LocalDate.now(KST) },
) : ViewModel() {
    private val _uiState = MutableStateFlow(DisplayUiState())
    val uiState: StateFlow<DisplayUiState> = _uiState.asStateFlow()

    private var latestLoadGeneration = 0
    private var selectedSort = PostSort.LATEST
    private var loadedDate: LocalDate? = null
    private var isEndThresholdReached = false
    private var isRevalidating = false
    private var hasBeenPresented = false
    private var nextPageJob: Job? = null
    private var nextMessageId = 0L
    private val displayCache = mutableMapOf<DisplayCacheKey, DisplayCacheEntry>()

    init {
        loadDisplay(date = initialDate)
    }

    fun moveToPreviousDate() {
        val state = _uiState.value
        if (state.content is DisplayContentState.Loading) return
        val selectedDate = state.selectedDate ?: return
        val targetDate = state.earliestDate
            ?.let { selectedDate.minusDays(1).coerceAtLeast(it) }
            ?: selectedDate.minusDays(1)
        if (targetDate == selectedDate) return

        startDateChange(targetDate, state)
    }

    fun moveToNextDate() {
        val state = _uiState.value
        if (state.content is DisplayContentState.Loading) return
        val selectedDate = state.selectedDate ?: return
        val latestDate = state.latestDate ?: return
        val targetDate = selectedDate.plusDays(1).coerceAtMost(latestDate)
        if (targetDate == selectedDate) return

        startDateChange(targetDate, state)
    }

    fun selectSort(sort: PostSort) {
        val state = _uiState.value
        val latestContent = state.content as? DisplayContentState.Latest ?: return
        if (sort == latestContent.selectedSort) return
        val selectedDate = state.selectedDate ?: return

        selectedSort = sort
        val cachedState = displayCache[DisplayCacheKey(selectedDate, sort)]?.state?.copy(
            latestDate = state.latestDate,
            earliestDate = state.earliestDate,
            pendingMessage = null,
        )
        val visibleState = cachedState ?: state.copy(
            content = latestContent.copy(selectedSort = sort),
            isLoadingNext = false,
        )
        _uiState.value = visibleState
        loadDisplay(
            date = selectedDate,
            previousState = cachedState ?: state,
            keepsContentVisible = true,
        )
    }

    fun retry() {
        loadDisplay(_uiState.value.selectedDate)
    }

    fun onResume() {
        if (!hasBeenPresented) {
            hasBeenPresented = true
            return
        }
        revalidate()
    }

    fun revalidate() {
        val state = _uiState.value
        if (state.content is DisplayContentState.Loading || isRevalidating) return

        val selectedDate = state.selectedDate ?: return
        loadDisplay(
            date = selectedDate,
            previousState = state,
            keepsContentVisible = true,
        )
    }

    fun onMessageShown(messageId: Long) {
        _uiState.update { state ->
            if (state.pendingMessage?.id == messageId) {
                state.copy(pendingMessage = null)
            } else {
                state
            }
        }
    }

    fun updateFeaturedPage(page: Int) {
        _uiState.update { state ->
            val archive = state.content as? DisplayContentState.Archive ?: return@update state
            val lastPage = archive.featuredPhotos.lastIndex
                .coerceAtLeast(0)
            state.copy(
                content = archive.copy(featuredPage = page.coerceIn(0, lastPage)),
            )
        }
    }

    fun updateEndThreshold(isReached: Boolean) {
        if (!isReached) {
            isEndThresholdReached = false
            return
        }
        if (isEndThresholdReached) return

        isEndThresholdReached = true
        loadNextPage()
    }

    private fun startDateChange(
        targetDate: LocalDate,
        previousState: DisplayUiState,
    ) {
        _uiState.update {
            it.copy(
                selectedDate = targetDate,
                content = DisplayContentState.Loading,
            )
        }
        loadDisplay(targetDate, previousState = previousState)
    }

    private fun loadDisplay(
        date: LocalDate?,
        sort: PostSort = selectedSort,
        previousState: DisplayUiState? = null,
        keepsContentVisible: Boolean = false,
    ) {
        val latestDate = dateProvider()
        val requestedDate = date ?: latestDate
        val requestedSort = if (requestedDate < latestDate) PostSort.POPULAR else sort
        val generation = ++latestLoadGeneration
        isRevalidating = keepsContentVisible
        nextPageJob?.cancel()
        nextPageJob = null
        isEndThresholdReached = false

        viewModelScope.launch {
            if (keepsContentVisible) {
                _uiState.update { it.copy(latestDate = latestDate, isLoadingNext = false) }
            } else {
                _uiState.update {
                    it.copy(
                        selectedDate = requestedDate,
                        latestDate = latestDate,
                        content = DisplayContentState.Loading,
                    )
                }
            }
            val result = try {
                repository.getPostContent(
                    HomeQuery(
                        date = requestedDate,
                        sort = requestedSort,
                        page = HomeQuery.FIRST_PAGE,
                    ),
                )
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (_: Exception) {
                HomeResult.Failure(HomeFailure.Network)
            }
            if (generation != latestLoadGeneration) return@launch

            when (result) {
                is HomeResult.Success -> applyFirstPage(
                    postContent = result.value,
                    latestDate = latestDate,
                    requestedSort = requestedSort,
                )

                is HomeResult.Failure -> if (previousState != null) {
                    val isPreviousDateRequest = result.reason == HomeFailure.TopicNotFound &&
                        requireNotNull(date) < requireNotNull(previousState.selectedDate)
                    selectedSort = (previousState.content as? DisplayContentState.Latest)
                        ?.selectedSort
                        ?: selectedSort
                    val pendingMessage = nextToast(DISPLAY_ERROR_MESSAGE)
                    _uiState.update {
                        it.copy(
                            selectedDate = previousState.selectedDate,
                            topic = previousState.topic,
                            content = previousState.content,
                            likedPhotoIds = previousState.likedPhotoIds,
                            currentPage = previousState.currentPage,
                            hasNext = previousState.hasNext,
                            randomSeed = previousState.randomSeed,
                            isLoadingNext = false,
                            earliestDate = if (isPreviousDateRequest) {
                                previousState.selectedDate
                            } else {
                                previousState.earliestDate
                            },
                            pendingMessage = pendingMessage,
                        )
                    }
                } else {
                    _uiState.update {
                        it.copy(
                            content = DisplayContentState.Error(
                                message = DISPLAY_ERROR_MESSAGE,
                            ),
                        )
                    }
                }
            }
            isRevalidating = false
        }
    }

    private fun applyFirstPage(
        postContent: PostContent,
        latestDate: LocalDate,
        requestedSort: PostSort,
    ) {
        val earliestDate = _uiState.value.earliestDate
        loadedDate = postContent.topicDate
        val cacheKey = DisplayCacheKey(postContent.topicDate, requestedSort)
        val cachedEntry = displayCache[cacheKey]
        val canReuseCachedTail = requestedSort != PostSort.RANDOM ||
            (
                !postContent.randomSeed.isNullOrBlank() &&
                    cachedEntry?.state?.randomSeed == postContent.randomSeed
            )
        val freshPhotoIds = postContent.photos.mapTo(mutableSetOf(), Post::id)
        val cachedFirstPagePhotoIds = cachedEntry
            ?.firstPagePhotoIds
            .orEmpty()
            .takeIf { canReuseCachedTail }
            .orEmpty()
        val cachedTail = if (canReuseCachedTail) {
            cachedEntry
                ?.state
                ?.photos
                .orEmpty()
                .filter { it.id !in cachedFirstPagePhotoIds && it.id !in freshPhotoIds }
        } else {
            emptyList()
        }
        val mergedPhotos = postContent.photos + cachedTail
        val cachedTailIds = cachedTail.mapTo(mutableSetOf(), Post::id)
        val mergedLikedPhotoIds = postContent.likedPhotoIds +
            cachedEntry
                ?.state
                ?.likedPhotoIds
                .orEmpty()
                .intersect(cachedTailIds)
        val content = if (postContent.topicDate < latestDate) {
            DisplayContentState.Archive(
                photos = mergedPhotos,
                featuredPhotos = mergedPhotos.take(FEATURED_PHOTO_COUNT),
            )
        } else {
            DisplayContentState.Latest(
                photos = mergedPhotos,
                selectedSort = selectedSort,
            )
        }
        val newState = DisplayUiState(
            selectedDate = postContent.topicDate,
            latestDate = latestDate,
            earliestDate = earliestDate,
            topic = postContent.topic,
            content = content,
            likedPhotoIds = mergedLikedPhotoIds,
            currentPage = cachedEntry
                ?.state
                ?.currentPage
                ?.takeIf { cachedTail.isNotEmpty() }
                ?: postContent.currentPage,
            hasNext = cachedEntry
                ?.state
                ?.hasNext
                ?.takeIf { cachedTail.isNotEmpty() }
                ?: postContent.hasNext,
            randomSeed = postContent.randomSeed,
        )
        _uiState.value = newState
        displayCache[cacheKey] = DisplayCacheEntry(
            state = newState,
            firstPagePhotoIds = freshPhotoIds,
        )
    }

    private fun loadNextPage() {
        val state = _uiState.value
        val date = loadedDate ?: return
        if (state.content !is DisplayContentState.Latest &&
            state.content !is DisplayContentState.Archive
        ) {
            return
        }
        if (isRevalidating || !state.hasNext || state.isLoadingNext || nextPageJob != null) return

        val sort = when (val content = state.content) {
            is DisplayContentState.Latest -> content.selectedSort
            is DisplayContentState.Archive -> PostSort.POPULAR
            else -> return
        }
        val randomSeed = state.randomSeed.takeIf { sort == PostSort.RANDOM }
        if (sort == PostSort.RANDOM && randomSeed.isNullOrBlank()) {
            _uiState.update { it.copy(hasNext = false) }
            return
        }

        val generation = latestLoadGeneration
        val query = HomeQuery(
            date = date,
            sort = sort,
            page = state.currentPage + 1,
            randomSeed = randomSeed,
        )
        _uiState.update { it.copy(isLoadingNext = true) }
        val job = viewModelScope.launch {
            val result = try {
                repository.getPostPage(query)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (_: Exception) {
                HomeResult.Failure(HomeFailure.Network)
            }
            if (generation != latestLoadGeneration) return@launch

            when (result) {
                is HomeResult.Success -> appendPage(result.value)
                is HomeResult.Failure -> _uiState.update { it.copy(isLoadingNext = false) }
            }
        }
        nextPageJob = job
        job.invokeOnCompletion {
            if (nextPageJob === job) nextPageJob = null
        }
    }

    private fun appendPage(page: PostPage) {
        _uiState.update { state ->
            val content = state.content
            val existingIds = when (content) {
                is DisplayContentState.Latest -> content.photos.mapTo(mutableSetOf(), Post::id)
                is DisplayContentState.Archive -> content.photos.mapTo(mutableSetOf(), Post::id)
                else -> return@update state.copy(isLoadingNext = false)
            }
            val newPhotos = page.photos.filter { existingIds.add(it.id) }
            val newPhotoIds = newPhotos.mapTo(mutableSetOf(), Post::id)
            val updatedContent = when (content) {
                is DisplayContentState.Latest -> content.copy(photos = content.photos + newPhotos)
                is DisplayContentState.Archive -> content.copy(photos = content.photos + newPhotos)
                else -> return@update state.copy(isLoadingNext = false)
            }
            state.copy(
                content = updatedContent,
                likedPhotoIds = (state.likedPhotoIds - newPhotoIds) +
                    page.likedPhotoIds.intersect(newPhotoIds),
                currentPage = page.currentPage,
                hasNext = page.hasNext,
                randomSeed = if (state.content is DisplayContentState.Latest &&
                    state.content.selectedSort == PostSort.RANDOM
                ) {
                    state.randomSeed ?: page.randomSeed
                } else {
                    null
                },
                isLoadingNext = false,
            )
        }
        cacheCurrentState()
    }

    private fun cacheCurrentState() {
        val state = _uiState.value
        val date = state.selectedDate ?: return
        val sort = when (val content = state.content) {
            is DisplayContentState.Latest -> content.selectedSort
            is DisplayContentState.Archive -> PostSort.POPULAR
            else -> return
        }
        val key = DisplayCacheKey(date, sort)
        displayCache[key] = DisplayCacheEntry(
            state = state.copy(pendingMessage = null),
            firstPagePhotoIds = displayCache[key]?.firstPagePhotoIds
                ?: state.photos.mapTo(mutableSetOf(), Post::id),
        )
    }

    companion object {
        fun factory(initialDate: LocalDate? = null) = viewModelFactory {
            initializer {
                val application = this[APPLICATION_KEY] as ChalkakApplication
                DisplayViewModel(
                    repository = application.appContainer.postRepository,
                    initialDate = initialDate,
                )
            }
        }

        val Factory = factory()

        private val KST: ZoneId = ZoneId.of("Asia/Seoul")
        private const val FEATURED_PHOTO_COUNT = 5
    }

    private fun nextToast(text: String): UiMessage.Toast = UiMessage.Toast(
        id = nextMessageId++,
        text = text,
    )
}

private data class DisplayCacheKey(
    val date: LocalDate,
    val sort: PostSort,
)

private data class DisplayCacheEntry(
    val state: DisplayUiState,
    val firstPagePhotoIds: Set<String>,
)

private val DisplayUiState.photos: List<Post>
    get() = when (val content = content) {
        is DisplayContentState.Latest -> content.photos
        is DisplayContentState.Archive -> content.photos
        else -> emptyList()
    }

private const val DISPLAY_ERROR_MESSAGE = "전시를 불러오지 못했어요"

File: client/iOS/chalkak/chalkak/Features/Display/DisplayViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class DisplayViewModel {
    typealias DateProvider = @MainActor @Sendable () -> Date
    typealias FirstPageHandler = @MainActor @Sendable (
        Date,
        DisplaySort
    ) async -> Result<DisplayContent, DisplayError>
    typealias NextPageHandler = @MainActor @Sendable (
        DisplayPageRequest
    ) async -> Result<DisplayPage, DisplayError>

    private(set) var viewState: DisplayViewState
    private(set) var event: DisplayEvent?

    private let initialDate: Date?
    private let dateProvider: DateProvider
    private let firstPageHandler: FirstPageHandler
    private let nextPageHandler: NextPageHandler
    private var selectedLatestSort: DisplaySort
    private var loadedTopicDate: Date?
    private var generation = 0
    private var isEndThresholdReached = false

    init(
        initialState: DisplayViewState? = nil,
        initialDate: Date? = nil,
        dateProvider: @escaping DateProvider = { Date() },
        firstPageHandler: @escaping FirstPageHandler = { _, _ in .failure(.generic) },
        nextPageHandler: @escaping NextPageHandler = { _ in .failure(.generic) }
    ) {
        let initialState = initialState ?? DisplayViewState()
        self.viewState = initialState
        self.initialDate = initialDate
        self.dateProvider = dateProvider
        self.firstPageHandler = firstPageHandler
        self.nextPageHandler = nextPageHandler
        self.selectedLatestSort = initialState.selectedSort
    }

    convenience init(
        initialDate: Date? = nil,
        dateProvider: @escaping DateProvider = { Date() },
        apiClient: DisplayAPIClient
    ) {
        self.init(
            initialDate: initialDate,
            dateProvider: dateProvider,
            firstPageHandler: { date, sort in
                do {
                    return .success(try await apiClient.fetchDisplay(date: date, sort: sort))
                } catch is CancellationError {
                    return .failure(.generic)
                } catch let error as DisplayAPIError {
                    return .failure(error.displayError)
                } catch {
                    return .failure(.generic)
                }
            },
            nextPageHandler: { request in
                do {
                    return .success(try await apiClient.fetchPage(request))
                } catch is CancellationError {
                    return .failure(.generic)
                } catch let error as DisplayAPIError {
                    return .failure(error.displayError)
                } catch {
                    return .failure(.generic)
                }
            }
        )
    }

    func load() async {
        let latestDate = Self.startOfDay(dateProvider())
        let requestedDate = Self.startOfDay(initialDate ?? latestDate)
        await loadFirstPage(
            date: requestedDate,
            latestDate: latestDate,
            preserves: nil
        )
    }

    func retry() async {
        let previous = viewState.hasLoadedContent ? viewState : nil
        let providerDate = Self.startOfDay(dateProvider())
        let latestDate = viewState.contentStatus == .loading
            ? providerDate
            : viewState.latestDate ?? providerDate
        let requestedDate = viewState.contentStatus == .loading
            ? initialDate ?? latestDate
            : viewState.selectedDate ?? initialDate ?? latestDate
        await loadFirstPage(date: requestedDate, latestDate: latestDate, preserves: previous)
    }

    func moveToPreviousDate() async {
        guard viewState.contentStatus != .loading,
              let selectedDate = viewState.selectedDate
        else { return }
        let previousDate = Self.calendar.date(byAdding: .day, value: -1, to: selectedDate)!
        let targetDate = viewState.earliestDate.map { max(previousDate, $0) } ?? previousDate
        guard targetDate != selectedDate else { return }

        await loadFirstPage(
            date: targetDate,
            latestDate: viewState.latestDate ?? Self.startOfDay(dateProvider()),
            preserves: viewState,
            isPreviousDateRequest: true
        )
    }

    func moveToNextDate() async {
        guard viewState.contentStatus != .loading,
              let selectedDate = viewState.selectedDate,
              let latestDate = viewState.latestDate
        else { return }
        let nextDate = Self.calendar.date(byAdding: .day, value: 1, to: selectedDate)!
        let targetDate = min(nextDate, latestDate)
        guard targetDate != selectedDate else { return }

        await loadFirstPage(
            date: targetDate,
            latestDate: latestDate,
            preserves: viewState
        )
    }

    func selectSort(_ sort: DisplaySort) async {
        guard viewState.contentStatus == .latest,
              sort != viewState.selectedSort,
              let selectedDate = viewState.selectedDate,
              let latestDate = viewState.latestDate
        else { return }

        selectedLatestSort = sort
        await loadFirstPage(
            date: selectedDate,
            latestDate: latestDate,
            preserves: viewState
        )
    }

    func updateFeaturedPage(_ page: Int) {
        guard viewState.contentStatus == .archive else { return }
        viewState.featuredPage = page.coerced(
            to: 0...max(0, viewState.featuredPhotos.count - 1)
        )
    }

    func didReachEndThreshold(_ isReached: Bool) async {
        guard isReached else {
            isEndThresholdReached = false
            return
        }
        guard !isEndThresholdReached else { return }
        isEndThresholdReached = true
        await loadNextPage()
    }

    func consumeEvent() {
        event = nil
    }

    private func loadFirstPage(
        date: Date,
        latestDate: Date,
        preserves previousState: DisplayViewState?,
        isPreviousDateRequest: Bool = false
    ) async {
        generation += 1
        let requestGeneration = generation
        isEndThresholdReached = false
        let requestedDate = Self.startOfDay(date)
        let requestSort: DisplaySort = requestedDate < latestDate ? .popular : selectedLatestSort

        viewState.selectedDate = requestedDate
        viewState.latestDate = latestDate
        viewState.contentStatus = .loading
        viewState.isLoadingNext = false
        viewState.transientError = nil

        let result = await firstPageHandler(requestedDate, requestSort)
        guard requestGeneration == generation else { return }

        switch result {
        case let .success(content):
            apply(content, latestDate: latestDate)
        case let .failure(error):
            loadedTopicDate = previousState.flatMap { _ in loadedTopicDate }
            if var restored = previousState {
                if isPreviousDateRequest, error == .topicNotFound {
                    restored.earliestDate = restored.selectedDate
                }
                selectedLatestSort = restored.contentStatus == .latest
                    ? restored.selectedSort
                    : selectedLatestSort
                restored.latestDate = latestDate
                restored.isLoadingNext = false
                restored.transientError = error
                viewState = restored
                event = .showFailure(error)
            } else {
                viewState = DisplayViewState(
                    contentStatus: .error(error),
                    selectedDate: requestedDate,
                    latestDate: latestDate,
                    selectedSort: selectedLatestSort
                )
            }
        }
    }

    private func apply(_ content: DisplayContent, latestDate: Date) {
        let canonicalDate = Self.startOfDay(content.topicDate)
        let isArchive = canonicalDate < latestDate
        loadedTopicDate = canonicalDate
        viewState = DisplayViewState(
            contentStatus: isArchive ? .archive : .latest,
            selectedDate: canonicalDate,
            latestDate: latestDate,
            earliestDate: viewState.earliestDate,
            topic: content.topic,
            selectedSort: isArchive ? .popular : selectedLatestSort,
            photos: content.page.photos,
            featuredPhotos: isArchive ? Array(content.page.photos.prefix(5)) : [],
            featuredPage: 0,
            currentPage: content.page.currentPage,
            hasNext: content.page.hasNext,
            randomSeed: isArchive ? nil : content.page.randomSeed,
            isLoadingNext: false
        )
    }

    private func loadNextPage() async {
        guard viewState.contentStatus == .latest || viewState.contentStatus == .archive,
              viewState.hasNext,
              !viewState.isLoadingNext,
              let loadedTopicDate
        else {
            isEndThresholdReached = false
            return
        }

        let sort: DisplaySort = viewState.contentStatus == .archive ? .popular : viewState.selectedSort
        let randomSeed = sort == .random ? viewState.randomSeed : nil
        if sort == .random, randomSeed?.isEmpty != false {
            viewState.hasNext = false
            isEndThresholdReached = false
            return
        }

        let requestGeneration = generation
        let request = DisplayPageRequest(
            topicDate: loadedTopicDate,
            sort: sort,
            page: viewState.currentPage + 1,
            randomSeed: randomSeed
        )
        viewState.isLoadingNext = true
        let result = await nextPageHandler(request)
        guard requestGeneration == generation else { return }

        switch result {
        case let .success(page):
            append(page, sort: sort)
        case let .failure(error):
            viewState.isLoadingNext = false
            viewState.transientError = error
            isEndThresholdReached = false
            event = .showFailure(error)
        }
    }

    private func append(_ page: DisplayPage, sort: DisplaySort) {
        var existingIDs = Set(viewState.photos.map(\.id))
        let newPhotos = page.photos.filter { existingIDs.insert($0.id).inserted }
        viewState.photos.append(contentsOf: newPhotos)
        viewState.currentPage = page.currentPage
        viewState.hasNext = page.hasNext
        viewState.randomSeed = sort == .random ? viewState.randomSeed ?? page.randomSeed : nil
        viewState.isLoadingNext = false
        isEndThresholdReached = false
    }

    private static var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "Asia/Seoul")!
        return calendar
    }

    private static func startOfDay(_ date: Date) -> Date {
        calendar.startOfDay(for: date)
    }
}

private extension DisplayViewState {
    var hasLoadedContent: Bool {
        contentStatus == .latest || contentStatus == .archive
    }
}

private extension Int {
    func coerced(to range: ClosedRange<Int>) -> Int {
        Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
    }
}
